bindings/rust/bench/src/harness.rs

```diff
@@ -12,14 +19,144 @@ pub enum Mode {
     Server,
 }
 
-pub trait TlsBenchHarness {
+// these parameters were the only ones readily usable for all three libaries:
+// s2n-tls, rustls, and openssl
+#[allow(non_camel_case_types)]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CipherSuite {
+    AES_128_GCM_SHA256,
+    AES_256_GCM_SHA384,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ECGroup {
+    SECP256R1,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct CryptoConfig {
+    pub cipher_suite: CipherSuite,
+    pub ec_group: ECGroup,
+}
+
+pub trait TlsBenchHarness: Sized {
+    /// Default harness
+    fn default() -> Result<Self, Box<dyn Error>> {
+        Self::new(&CryptoConfig {
+            cipher_suite: CipherSuite::AES_128_GCM_SHA256,
+            ec_group: ECGroup::SECP256R1,
+        })
+    }
+
     /// Initialize buffers, configs, and connections (pre-handshake)
-    fn new() -> Self;
+    fn new(crypto_config: &CryptoConfig) -> Result<Self, Box<dyn Error>>;
 
     /// Run handshake on initialized connection
-    /// Returns error if handshake already happened
-    fn handshake(&mut self) -> Result<(), &str>;
+    /// Returns error if handshake has already completed
+    fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
     /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
+
+    /// Get negotiated cipher suite
+    ///
+    /// Returns `Err` if handshake has not completed
+    fn get_negotiated_cipher_suite(&self) -> Result<CipherSuite, Box<dyn Error>>;
+
+    /// Get whether or negotiated version is TLS1.3
+    ///
+    /// Retuns `Err` if handshake has not completed
+    fn negotiated_tls13(&self) -> Result<bool, Box<dyn Error>>;
```

Comment:
I wonder if these methods would be simpler if they just panicked instead of returning a result? I think the results make sense for the other functions to avoid copious amounts of `unwrap` but these seem like it might be easier to deal with guaranteed results.

Reply:
For both Rustls and s2n-tls, `new()` processes the most amount of results and so returning a result is especially useful. I wanted to bubble up the results in the rest of the functions as well just for consistency: it feels weird to me if `new()` never panics but `handshake()` might.

Reply:
Sorry, by "these methods" I was specifically referring to
- get_negotiated_cipher_suite
- negotiated_tls13

Reply:
Would an Option be better? Panicking might be overreacting, but an Option implies that there could sometimes be a value and sometimes not, which is the case.

Reply:
Talked offline, choosing to panic instead of returning result for simplicity, knowing that the methods shouldn't panic in a testing environment anyways.